cherab/core/laser/scattering.pyx

```diff
@@ -0,0 +1,199 @@
+        self._laser.notifier.add(self._laser_changed)
+
+        self._laser_changed()
+
+    @property
+    def plasma(self):
+        return self._plasma
+
+    @plasma.setter
+    def plasma(self, value):
+
+        if not isinstance(value, Plasma):
+            raise TypeError("Value has to be of type Plasma, but {0} passed".format(type(value)))
+
+        self._plasma = value
+
+    @property
+    def laser_model(self):
+        return self._laser_model
+
+    @laser_model.setter
+    def laser_model(self, LaserModel value):
+
+        self._laser_model = value
+
+    def set_laser_spectrum(self, LaserSpectrum value):
+        #unregister callback
+        if self._laser_spectrum is not None:
+            self._laser_spectrum._notifier.remove(self._laser_spectrum_changed)
+
+        self._laser_spectrum = value
+        self._laser_spectrum._notifier.add(self._laser_spectrum_changed)
+        self._laser_spectrum_changed()
+
+    def _laser_spectrum_changed(self):
+        self._laser_bins = self._laser_spectrum._bins
+        self._laser_wavelength_mv = self._laser_spectrum._wavelengths_mv
+        self._laser_power_mv = self._laser_spectrum._power_mv
+        
+cdef class SeldenMatobaThomsonSpectrum(ScatteringModel):
+
+    def __init__(self, Laser laser=None, LaserModel laser_models=None, Plasma plasma=None):
+        # from article: 2 * alpha = m_e * c **2 /(k * T_e), here rewritten for Te in eV
+        self._CONST_ALPHA = ELECTRON_REST_MASS * SPEED_OF_LIGHT ** 2 / ( 2 * ELEMENTARY_CHARGE)
+        if laser is not None:
+            self.laser = laser
+        else:
+            if laser_models is not None:
+                self.laser_model = laser_models
+            if plasma is not None:
+                self.plasma = plasma
+
+
+
+    @cython.cdivision(True)
+    cdef double seldenmatoba_spectral_shape(self, double epsilon, double cos_theta, double alpha):
+
+        cdef:
+            double c, a, b
+
+        c = sqrt(alpha / M_PI) * (1 - 15. / (16. * alpha) + 345. / (512. * alpha ** 2))
+        a = (1 + epsilon) ** 3 * sqrt(2 * (1 - cos_theta) * (1 + epsilon) + epsilon ** 2)
+        b = sqrt(1 + epsilon ** 2 / (2 * (1 - cos_theta) * (1 + epsilon))) - 1
+        return c / a * exp(-2 * alpha * b)
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    cpdef Spectrum emission(self, Point3D position_plasma, Point3D position_laser, Vector3D direction_observation,
+                            Spectrum spectrum):
+
+        cdef:
+            double ne, te, laser_power_density, angle_pointing, angle_polarization, laser_wavelength
+            Vector3D pointing_vector, polarization_vector
+            Spectrum laser_spectrum
+            Py_ssize_t index
+        
+        ne = self._plasma.get_electron_distribution().density(position_plasma.x, position_plasma.y, position_plasma.z)
+        te = self._plasma.get_electron_distribution().effective_temperature(position_plasma.x, position_plasma.y, position_plasma.z)
```

Comment:
Rather than make two calls to `self._plasma.get_electron_distribution()`, can you store the distribution in a local variable once and then access that each time?

Reply:
you mean something like:
``` python
e_distribution = self._plasma.get_electron_distribution()
ne = distribution.density(position_plasma.x, position_plasma.y, position_plasma.z)
te = distribution.effective_temperature(position_plasma.x, position_plasma.y, position_plasma.z) 
```
?

I know that look-ups of functions can take time, is this the reason?Or is it just readability?

Reply:
Yes, that's what I had in mind. I doubt it will make a significant difference to performance in this case, assuming the function is just a simple getter. But it does help readability, as it means this line of code is only doing one thing (getting the density of the distribution) rather than multiple things (first get the electron distribution of the plasma object, then get the density from that).